File: packages/shared/src/types/commit.ts
export interface Commit {
  sha: string;
  message: string;
  author: Author;
  committer: Author;
  date: Date;
  parents: string[];
  branches: string[];
  tags: string[];
  isStash: boolean;
}

export interface Author {
  name: string;
  email: string;
  date?: Date;
}

export interface CommitDetails extends Commit {
  stats: CommitStats;
  files: CommitFile[];
}

export interface CommitStats {
  additions: number;
  deletions: number;
  total: number;
  filesChanged: number;
}

export interface CommitFile {
  path: string;
  oldPath?: string;
  status: FileStatus;
  additions: number;
  deletions: number;
  diff?: string;
}

export type FileStatus = 'added' | 'modified' | 'deleted' | 'renamed' | 'copied';

export interface LogOptions {
  branch?: string;
  maxCount?: number;
  skip?: number;
  author?: string;
  since?: Date;
  until?: Date;
  path?: string;
}

File: packages/backend.net/Domain/Commit.cs
namespace ForkWeb.Backend.Domain;

public record Commit(
    string Sha,
    string Message,
    Author Author,
    Author Committer,
    DateTime Date,
    List<string> Parents,
    List<string>? Branches = null,
    List<string>? Tags = null
);

public record Author(
    string Name,
    string Email,
    DateTime? Date = null
);

public record CommitDetails(
    string Sha,
    string Message,
    Author Author,
    Author Committer,
    DateTime Date,
    List<string> Parents,
    CommitStats Stats,
    List<CommitFile> Files,
    List<string>? Branches = null,
    List<string>? Tags = null
);

public record CommitStats(
    int Additions,
    int Deletions,
    int Total,
    int FilesChanged
);

public record CommitFile(
    string Path,
    string? OldPath,
    FileStatus Status,
    int Additions,
    int Deletions,
    string? Diff = null
);

public enum FileStatus
{
    Added,
    Modified,
    Deleted,
    Renamed,
    Copied
}

public record LogOptions(
    string? Branch = null,
    int? MaxCount = null,
    int? Skip = null,
    string? Author = null,
    DateTime? Since = null,
    DateTime? Until = null,
    string? Path = null
);
